include/zephyr/drivers/stepper.h

```diff
@@ -0,0 +1,425 @@
+ *
+ * (Internal use only.)
+ */
+
+/**
+ * @brief enable or disable the stepper motor controller.
+ *
+ * @see stepper_enable() for details.
+ */
+typedef int (*stepper_enable_t)(const struct device *dev, bool enable);
+
+/**
+ * @brief Move the stepper motor by a given number of micro_steps.
+ *
+ * @see stepper_move() for details.
+ */
+typedef int (*stepper_move_t)(const struct device *dev, int32_t micro_steps,
+				  struct k_poll_signal *async);
+
+/**
+ * @brief Set the max velocity in micro_steps per seconds.
+ *
+ * @see stepper_set_max_velocity() for details.
+ */
+typedef int (*stepper_set_max_velocity_t)(const struct device *dev,
+					      uint32_t micro_steps_per_second);
+
+/**
+ * @brief Set the micro-step resolution
+ *
+ * @see stepper_set_micro_step_res() for details.
+ */
+typedef int (*stepper_set_micro_step_res_t)(const struct device *dev,
+						enum micro_step_resolution resolution);
+
+/**
+ * @brief Get the micro-step resolution
+ *
+ * @see stepper_get_micro_step_res() for details.
+ */
+typedef int (*stepper_get_micro_step_res_t)(const struct device *dev,
+						enum micro_step_resolution *resolution);
+/**
+ * @brief Set the actual a.k.a reference position of the stepper
+ *
+ * @see stepper_set_actual_position() for details.
+ */
+typedef int (*stepper_set_actual_position_t)(const struct device *dev, const int32_t value);
+
+/**
+ * @brief Get the actual a.k.a reference position of the stepper
+ *
+ * @see stepper_get_actual_position() for details.
+ */
+typedef int (*stepper_get_actual_position_t)(const struct device *dev, int32_t *value);
+
+/**
+ * @brief Set the absolute target position of the stepper
+ *
+ * @see stepper_set_target_position() for details.
+ */
+typedef int (*stepper_set_target_position_t)(const struct device *dev, const int32_t value,
+						 struct k_poll_signal *async);
+
+/**
+ * @brief Is the target position fo the stepper reached
+ *
+ * @see stepper_is_moving() for details.
+ */
+typedef int (*stepper_is_moving_t)(const struct device *dev, bool *is_moving);
+
+/**
+ * @brief Enable constant velocity mode for the stepper with a given velocity
+ *
+ * @see stepper_enable_constant_velocity_mode() for details.
+ */
+typedef int (*stepper_enable_constant_velocity_mode_t)(const struct device *dev,
+							   const enum stepper_direction direction,
+							   const uint32_t value);
+
+/**
+ * @brief Stepper Motor Controller API
+ */
+__subsystem struct stepper_api {
+	stepper_enable_t enable;
+	stepper_move_t move;
+	stepper_set_max_velocity_t set_max_velocity;
+	stepper_set_micro_step_res_t set_micro_step_res;
+	stepper_get_micro_step_res_t get_micro_step_res;
+	stepper_set_actual_position_t set_actual_position;
+	stepper_get_actual_position_t get_actual_position;
+	stepper_set_target_position_t set_target_position;
+	stepper_is_moving_t is_moving;
+	stepper_enable_constant_velocity_mode_t enable_constant_velocity_mode;
```

Comment:
where is the velocity mode example in the GPIO example – it is possible but not implemented!